### PR TITLE
init_deco and CNS calculation - correctly identify previous dives and report overlapping dives

### PR DESCRIPTION
**Pull request by sfuchs79**
### Describe the pull request:

- [x] Bug fix
- [ ] Functional change
- [x] New feature
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

When changing the date/time of a dive in the planner the dive may end
up in a totaly new position in respect to date/time of other dives in
dive list table. It can be moved to the past or the future before or after
other existing dives. It also could overlap with an existing dive.

This change enables identification of a new "virtual" dive list position
and based on this starts looking for previous dives.
Then it (as before the change) does init the deco/CNS calculation with any
applicable previous dive and surface interval.
If some of these applicable dives overlap it returns a neg. surface time
which is then used in the planner notes to prohibit display of results.
### Additional information:

This is work in progress and needs further review/testing.
I for the moment put a lot of debug output which would be removed or turned off by default lateron.

### CODE REVIEW

**Comment by dirkhh** [comment]
YAY @neolit123  - mention-bot seems to be back!

**Comment by neolit123** [comment]
@dirkhh 4 days ago someone posted in their issue tracker that the bot still isn't working.
at the same time, i don't see any relevant changes, that might have fixed that in their repository, so this is kind of odd.

i will ping them about it suddenly working here. but to my understanding there is no one to maintain this from the Facebook team.

**Comment by neolit123** [comment]
@dirkhh 
got a response from a developer that the bot account was banned for some reason and that now it's fixed and it should work fine!

**Comment by dirkhh** [comment]
@neolit123 thanks for letting us know! Now let's find time to look at the PR (sorry @sfuchs79 for the chatter about the unrelated mention-bot here...)

**Comment by sfuchs79** [comment]
I was so close to complain ;-)

Highlighting my own topic again - I pushed another commit with a second similar change for CNS calculation.

File: core/divelist.c [hunk 1]
```diff
@@ -259,6 +236,100 @@ static int calculate_cns(struct dive *dive)
...
+			break;
+		i++;
+	}
+	/* Look at previous dive in dive list table and correct i when needed */
+	while (i > 0) {
+		struct dive *pdive = get_dive(i - 1);
+		if (!pdive || pdive->when < dive->when)
+			break;
+		i--;
+	}
```
Comment (neolit123) [inline core/divelist.c:277]: given that `sort_table()` uses `when` from the `struct dive`, do you think that the second iteration is still needed? technically if the list is sorted, you would know where `dive` fits chronologically.
is the list not sorted at this point in time?

File: core/divelist.c [hunk 2]
```diff
@@ -367,94 +438,128 @@ int get_divesite_idx(struct dive_site *ds)
...
 		surface_pressure = get_surface_pressure_in_mbar(pdive, true) / 1000.0;
+		/* Is it the first dive we add? */
 		if (!deco_init) {
+			printf("Init deco\n");
 			clear_deco(surface_pressure);
 			deco_init = true;
-#if DECO_CALC_DEBUG & 2
+			printf("Tissues after init:\n");
 			dump_tissues();
-#endif
```
Comment (neolit123) [inline core/divelist.c]: we probably don't want to dump the tissues all the time by removing the `#if DECO_CALC_DEBUG & 2`. i'm assuming you did this so that you can monitor the changes without supplying the macro definition to the compiler?

File: core/divelist.c [hunk 3]
```diff
@@ -367,94 +438,128 @@ int get_divesite_idx(struct dive_site *ds)
...
+		}
+		add_segment(surface_pressure, &air, surface_time, 0, dive, prefs.decosac);
+		printf("Tissues after surface intervall of %d:%02u:\n", FRACTION(surface_time, 60));
 		dump_tissues();
-#endif
 	}
+
 	// I do not dare to remove this call. We don't need the result but it might have side effects. Bummer.
-	tissue_tolerance_calc(dive, surface_pressure);
+	//tissue_tolerance_calc(dive, surface_pressure);
```
Comment (neolit123) [inline core/divelist.c]: judging from the comment above the `tissue_tolerance_calc(...)` call, can you explain the reasoning behind commenting it out? :-)

this might go in a separate patch, i would say.

File: qt-models/diveplannermodel.cpp [hunk 4]
```diff
@@ -1039,6 +1039,7 @@ void DivePlannerPointsModel::createPlan(bool replanCopy)
 		copy_dive(&displayed_dive, current_dive);
 	}
 	mark_divelist_changed(true);
+	sort_table(&dive_table);
```
Comment (neolit123) [inline qt-models/diveplannermodel.cpp:1079]: hmm, so here we have a table sort. is this happening *after* the `init_decompression()` and `calculate_cns_dive()` calls?

File: core/plannernotes.c [hunk 5]
```diff
@@ -85,22 +85,30 @@ void add_plan_to_notes(struct diveplan *diveplan, struct dive *dive, bool show_d
...
+	} else {
+		len += snprintf(buffer + len, sz_buffer - len, "<div><b>%s (%s) %s %d:%02d) %s %s<br>",
+				translate("gettextFromC", "Subsurface"),
+				subsurface_canonical_version(),
+				translate("gettextFromC", "dive plan</b> (surface interval "),
+				FRACTION(diveplan->surface_interval / 60, 60),
+				translate("gettextFromC", "created on"),
+				get_current_date());
 	}
-
```
Comment (neolit123) [inline core/plannernotes.c]: no need for this whitespace change - line removal.

File: core/divelist.c [hunk 6]
```diff
@@ -259,6 +236,100 @@ static int calculate_cns(struct dive *dive)
...
+		}
+		/* Don't add the copy of the dive itself */
+		if (i == divenr) {
+			printf("No - copy of dive\n");
+			continue;
+		}
+		printf("Yes\n");
+
+		/* CNS reduced with 90min halftime during surface interval */
+		if (last_endtime) cns = cns * 1 / pow(2, (pdive->when - last_endtime) / (90.0 * 60.0));
```
[the remark this answers is not in the document]
Comment (neolit123) [reply core/divelist.c]: probably best to simplify the `cns * 1 / x` to `cns / x` for readabilit, too.

**Review by neolit123 — COMMENTED** [review]
i've added some generic comments, but i think someone else needs to look into the logic, too.

File: qt-models/diveplannermodel.cpp (hunk 4 above)
Comment (sfuchs79) [reply qt-models/diveplannermodel.cpp:1079]: Yes, in the planner case this is happening after. It is happening at the moment when the user finally decides to save the planned dive to the dive list.

File: core/divelist.c (hunk 3 above)
Comment (sfuchs79) [reply core/divelist.c]: My mistake! I simply was curious if s.th. is going to explode if I remove this ;-)
BTW: The comment is correct: There is really at least one strange side effect when removing the call so we better keep it.

File: core/divelist.c (hunk 2 above)
Comment (sfuchs79) [reply core/divelist.c]: Yes, for the final version I will for sure switch off the debug output by default again. I maybe need some hint how to do it: Remove it completely or add the #if DECO_CALC_DEBUG & 2 everywhere again.

File: core/divelist.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (sfuchs79) [reply core/divelist.c]: Yes, will fix all these coding style items.

File: core/divelist.c (hunk 1 above)
Comment (sfuchs79) [reply core/divelist.c:277]: In the planner when you plan a new dive or replan an existing planned dive you will always have a dive which is "floating". What I mean is that the dive may have a dive table number but this number can be incorrect because the user can move the date of the dive around. At this moment the list is not re-sorted all the time. Re-sorting is done not before exiting the planner.
At the same moment we need to identify potential previous dives for this "floating" dive. This is why we need to identify something like a temporary dive table number.

**Comment by dirkhh** [comment]
@atdotde - what about this one?
